src/lib/ethers/zksync/wrapper.ts

```diff
@@ -0,0 +1,509 @@
+import { types, Provider, Wallet } from 'zksync-ethers'
+import { logger, SocketParams } from '../../Logger'
+
+interface TransactionParams {
+  from?: string
+  to: string
+  gas?: string
+  gasPrice?: string
+  value?: string
+  data: string
+  nonce?: string
+  type?: number
+  maxFeePerGas?: string
+  maxPriorityFeePerGas?: string
+}
+
+/**
+ * Wraps the `ether` wallet / signer abstraction so it's compatible with the wallet middleware of
+ * `eth-json-rpc-middleware`.
+ */
+class WalletWrapper {
+  chainId: number
+  defaultGasPrice!: number
+  defaultGasLimit!: number
+  estimateGasLimit: boolean
+  estimateGasPrice: boolean
+  ethGasPriceFactor: boolean
+  forceEIP155: boolean
+  gasPriceFactor!: number
+  gasLimitFactor!: number
+  interleaveBlocks: number
+  lastKnownBlock: number
+  provider: Provider
+  wallets: Wallet[]
+
+  constructor (
+    interleave_blocks: number,
+    gas_price: number,
+    gas_limit: number,
+    estimate_gas_limit: boolean,
+    estimate_gas_price: boolean,
+    gas_price_factor: number,
+    gas_limit_factor: number,
+    force_eip_155: boolean,
+    eth_gas_price_factor: boolean
+  ) {
+    this.defaultGasPrice = gas_price
+    this.defaultGasLimit = gas_limit
+    this.estimateGasLimit = estimate_gas_limit
+    this.estimateGasPrice = estimate_gas_price
+    this.ethGasPriceFactor = eth_gas_price_factor
+    this.forceEIP155 = force_eip_155
+    this.gasPriceFactor = gas_price_factor
+    this.gasLimitFactor = gas_limit_factor
+    this.interleaveBlocks = interleave_blocks
+    this.lastKnownBlock = 0
+    this.wallets = []
+  }
+
+  /**
+   * Populate essential transaction parameters, self-estimating gas price and/or gas limit if required.
+   * @param socket Socket parms where the RPC call is coming from.
+   * @param params Input params, to be validated and completed, if necessary.
+   * @returns
+   */
+  async composeTransaction (
+    socket: SocketParams,
+    params: TransactionParams
+  ): Promise<types.TransactionRequest> {
+    // Compose actual transaction:
+    let tx: types.TransactionRequest = {
+      from: params.from, // todo: ? utils.getA.getAddress(params.from) : undefined,
+      to: params.to,
+      value: params.value,
+      data: params.data,
+      nonce: params.nonce
+    }
+    if (this.forceEIP155) {
+      tx = {
+        ...tx,
+        chainId: this.provider.network.chainId
+      }
+    }
+    if (tx.from) {
+      logger.verbose({ socket, message: `> From:      ${tx.from}` })
+    }
+    logger.verbose({ socket, message: `> To:        ${tx.to || '(deploy)'}` })
+    logger.verbose({
+      socket,
+      message: `> Data:      ${
+        tx.data ? tx.data.toString().substring(0, 10) + '...' : '(transfer)'
+      }`
+    })
+    logger.verbose({ socket, message: `> Value:     ${tx.value || 0} wei` })
+    if (this.forceEIP155) {
+      logger.verbose({ socket, message: `> ChainId:   ${tx.chainId}` })
+    }
+
+    // // Complete tx type, if necessary:
```

Comment:
Is this necessary?

Reply:
Tracing chain id, yes. Commented block should rather be removed.